### ISSUE TITLE
Possible parse error with anonymous functions (laravel)

### ISSUE THREAD

**Issue by nevadascout** [HvyIndustries/crane#110]
Sometimes at the migration files (`database\migrations`) and sometimes at the seeders (`database\seeds`) in this test case, its a Laravel boilerplate project.

```
[INFO] Checking the current version of Crane
[INFO] Watching these files: {**/*.php}
[INFO] Preparing to parse 5387 PHP source files...
[INFO] Processing 0 stubs from C:\Users\ciruz\AppData\Roaming/Crane/phpstubs
[INFO] Parsed file 1 of 5387 : c:\vagrant\foobar\server.php
[INFO] Parsed file 2 of 5387 : c:\vagrant\foobar\app\helpers.php
[INFO] Parsed file 3 of 5387 : c:\vagrant\foobar\bootstrap\app.php
[INFO] Parsed file 4 of 5387 : c:\vagrant\foobar\bootstrap\autoload.php
[INFO] Parsed file 5 of 5387 : c:\vagrant\foobar\config\access.php
[INFO] Parsed file 6 of 5387 : c:\vagrant\foobar\config\app.php
[INFO] Parsed file 7 of 5387 : c:\vagrant\foobar\config\backend.php
[INFO] Parsed file 8 of 5387 : c:\vagrant\foobar\config\analytics.php
[INFO] Parsed file 9 of 5387 : c:\vagrant\foobar\config\auth.php
[INFO] Parsed file 10 of 5387 : c:\vagrant\foobar\config\cache.php
[INFO] Parsed file 11 of 5387 : c:\vagrant\foobar\config\breadcrumbs.php
[INFO] Parsed file 12 of 5387 : c:\vagrant\foobar\config\broadcasting.php
[INFO] Parsed file 13 of 5387 : c:\vagrant\foobar\config\database.php
[INFO] Parsed file 14 of 5387 : c:\vagrant\foobar\config\debugbar.php
[INFO] Parsed file 15 of 5387 : c:\vagrant\foobar\config\compile.php
[INFO] Parsed file 16 of 5387 : c:\vagrant\foobar\config\filesystems.php
[INFO] Parsed file 17 of 5387 : c:\vagrant\foobar\config\gravatar.php
[INFO] Parsed file 18 of 5387 : c:\vagrant\foobar\config\locale.php
[INFO] Parsed file 19 of 5387 : c:\vagrant\foobar\config\javascript.php
[INFO] Parsed file 20 of 5387 : c:\vagrant\foobar\config\mail.php
[INFO] Parsed file 21 of 5387 : c:\vagrant\foobar\config\log-viewer.php
[INFO] Parsed file 22 of 5387 : c:\vagrant\foobar\config\laravel-backup.php
[INFO] Parsed file 23 of 5387 : c:\vagrant\foobar\config\services.php
[INFO] Parsed file 24 of 5387 : c:\vagrant\foobar\config\session.php
[INFO] Parsed file 25 of 5387 : c:\vagrant\foobar\config\queue.php
[INFO] Parsed file 26 of 5387 : c:\vagrant\foobar\config\view.php
[INFO] Parsed file 27 of 5387 : c:\vagrant\foobar\public\index.php
[INFO] Parsed file 28 of 5387 : c:\vagrant\foobar\tests\TestCase.php
[INFO] Parsed file 29 of 5387 : c:\vagrant\foobar\vendor\autoload.php
[INFO] Parsed file 30 of 5387 : c:\vagrant\foobar\app\Console\Kernel.php
[INFO] Parsed file 31 of 5387 : c:\vagrant\foobar\app\Http\breadcrumbs.php
[INFO] Parsed file 32 of 5387 : c:\vagrant\foobar\app\Events\Event.php
[INFO] Parsed file 33 of 5387 : c:\vagrant\foobar\app\Http\Kernel.php
[INFO] Parsed file 34 of 5387 : c:\vagrant\foobar\app\Exceptions\GeneralException.php
[INFO] Parsed file 35 of 5387 : c:\vagrant\foobar\app\Http\routes.php
[INFO] Parsed file 36 of 5387 : c:\vagrant\foobar\app\Exceptions\Handler.php
[INFO] Parsed file 37 of 5387 : c:\vagrant\foobar\app\Models\Laender.php
[INFO] Parsed file 38 of 5387 : c:\vagrant\foobar\app\Providers\AppServiceProvider.php
[INFO] Parsed file 39 of 5387 : c:\vagrant\foobar\app\Jobs\Job.php
[INFO] Parsed file 40 of 5387 : c:\vagrant\foobar\app\Providers\EventServiceProvider.php
[INFO] Parsed file 41 of 5387 : c:\vagrant\foobar\app\Providers\AuthServiceProvider.php
[INFO] Parsed file 42 of 5387 : c:\vagrant\foobar\app\Providers\AccessServiceProvider.php
[INFO] Parsed file 43 of 5387 : c:\vagrant\foobar\app\Providers\RouteServiceProvider.php
[INFO] Parsed file 44 of 5387 : c:\vagrant\foobar\app\Providers\MacroServiceProvider.php
[INFO] Parsed file 45 of 5387 : c:\vagrant\foobar\bootstrap\cache\services.php
[INFO] Parsed file 46 of 5387 : c:\vagrant\foobar\database\factories\ModelFactory.php
[INFO] Parsed file 47 of 5387 : c:\vagrant\foobar\database\migrations\2014_10_12_100000_create_password_resets_table.php
[INFO] Parsed file 48 of 5387 : c:\vagrant\foobar\database\migrations\2015_12_28_171741_create_social_logins_table.php
```

It stucks at this file:

https://github.com/rappasoft/laravel-5-boilerplate/blob/master/database/migrations/2015_12_28_171741_create_social_logins_table.php

But if I start it again, it will stuck at another file (migrations or seeds)

**Comment by nevadascout** [comment]
It's very possible this is caused by the anonymous functions in those scripts. I'll do some investigation to find out more

**Comment by TheColorRed** [comment]
My seeders and migrations are processing just fine...

**Comment by ciruz** [comment]
I will give it a try with the boilerplate on another system.

**Comment by ciruz** [comment]
Made some tests on another system (with two php versions 5.6.16 and 7.0.0).

I tried a fresh Laravel project and a fresh [Laravel boilerplate project](https://github.com/rappasoft/laravel-5-boilerplate) and I didnt had this indexing problem here.

**Comment by ciruz** [comment]
A coworker tried it with a Zend project, got stuck at this:

```
[INFO] Parsed file 834 of 2882 : d:\foobar\library\Base\Pdf\Tcpdf\tcpdf_autoconfig.php
[INFO] Parsed file 835 of 2882 : d:\foobar\library\Base\Pdf\Tcpdf\tcpdf_barcodes_1d.php
[INFO] Parsed file 836 of 2882 : d:\foobar\library\Base\Pdf\Tcpdf\tcpdf_barcodes_2d.php
```

Last parsed file: https://github.com/contao-archive/tcpdf/blob/master/tcpdf_barcodes_2d.php
Next file would be: https://github.com/contao-archive/tcpdf/blob/master/tcpdf_import.php

**Comment by TheColorRed** [comment]
Of those three files, I was able to parse all of them on my computer and get auto-completion suggestions.

Might I ask how much RAM does your computer have? Could it be you are running out of memory to store the tree?

**Comment by ciruz** [comment]
16 GB

**Comment by marcw** [comment]
My analysis always stops ~65%. I currently have 14942 files to process in my project.  

Here are the files where it stops for the last 3 runs I tried: 
- [`vendor/doctrine/orm/lib/Doctrine/ORM/Persisters/Entity/EntityPersister.php`](https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/Persisters/Entity/EntityPersister.php)
- [`vendor/doctrine/orm/lib/Doctrine/ORM/Query/Exec/SingleSelectExecutor.php`](https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/Query/Exec/SingleSelectExecutor.php)
- [`vendor/friendsofphp/php-cs-fixer/Symfony/CS/Tests/Fixer/Symfony/PhpdocScalarFixerTest.php`](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/1.11/Symfony/CS/Tests/Fixer/Symfony/PhpdocScalarFixerTest.php)

Memory doesn't seem to be an issue as you can see:

![screenshot 2016-06-30 09 22 36](https://cloud.githubusercontent.com/assets/160332/16480177/f38ba9fc-3ea4-11e6-9450-7d3a10c6ff76.png)

**Comment by TheColorRed** [comment]
@marcw and @ciruz I assume you are both using Composer, would it be possible get a copy of your `composer.json` file? If we are able to duplicate your dependencies, we might be able to get to the root of the issue faster.

Thanks!

**Comment by marcw** [comment]
@TheColorRed You'll find it here: https://gist.github.com/marcw/3834cbe4099c7002497b60a20c5f1cd1

**Comment by ciruz** [comment]
I just tried it with VS Code Insiders and had no problems now. :open_mouth: 

My composer.json file was basically the same file from the boilerplate.

```json
{
    "name": "rappasoft/laravel-5-boilerplate",
    "description": "The Laravel Framework Boilerplate Project.",
    "keywords": ["framework", "laravel", "boilerplate"],
    "license": "MIT",
    "type": "project",
    "require": {
        "php": ">=5.5.9",
        "laravel/framework": "5.2.*",
        "laravelcollective/html": "5.2.*",
        "laracasts/utilities": "~2.0",
        "laravel/socialite": "^2.0",
        "creativeorange/gravatar": "~1.0",
        "davejamesmiller/laravel-breadcrumbs": "^3.0",
        "hieu-le/active": "~2.0",
        "arcanedev/log-viewer": "~3.0",
        "spatie/laravel-backup": "^3.0",
        "cviebrock/eloquent-sluggable": "^3.1"
    },
    "require-dev": {
        "fzaninotto/faker": "~1.4",
        "mockery/mockery": "0.9.*",
        "phpunit/phpunit": "~4.0",
        "symfony/css-selector": "2.8.*|3.0.*",
        "symfony/dom-crawler": "2.8.*|3.0.*",
        "barryvdh/laravel-debugbar": "^2.1",
        "laracasts/generators": "^1.1"
    },
    "autoload": {
        "classmap": [
            "database"
        ],
        "psr-4": {
            "App\\": "app/"
        },
        "files": [
            "app/helpers.php"
        ]
    },
    "autoload-dev": {
        "classmap": [
            "tests/TestCase.php"
        ]
    },
    "scripts": {
        "post-root-package-install": [
            "php -r \"copy('.env.example', '.env');\""
        ],
        "post-create-project-cmd": [
            "php artisan key:generate"
        ],
        "post-install-cmd": [
            "Illuminate\\Foundation\\ComposerScripts::postInstall",
            "php artisan optimize"
        ],
        "post-update-cmd": [
            "Illuminate\\Foundation\\ComposerScripts::postUpdate",
            "php artisan optimize"
        ]
    },
    "config": {
        "preferred-install": "dist"
    }
}
```

**Comment by TheColorRed** [comment]
Thanks guys! I will take a look at these when I have a chance!

**Comment by TheColorRed** [comment]
@marcw I loaded your composer.json file, and opened the project up. I was able to parse the entire project. They did come out with a new version of vscode, have you tried running it on the new version? That is the version I ran your project with.

**Comment by marcw** [comment]
@TheColorRed You are right. I updated VsCode and now it works fine. Many thanks :-)

**Comment by nazwa** [comment]
Did you guys ever find out what may have been causing this? I tried crane a few weeks ago and it was getting stuck, and now with the latest VS update the same thing still happens. Seems to be totally random and hangs the entire app.

**Comment by nevadascout** [comment]
@nazwa I'm afraid not. Are you able to give us some more information about which file is causing the freezing?

**Comment by nazwa** [comment]
That's the problem, it seems to be totally random every time. Whenever it happens, soon after VS shows a dialog about app not responding and asks if I want to wait or reload. 
Could it be the indexing loop blocking the message queue or not responding to heartbeats?

**Comment by nevadascout** [comment]
Fixed sed in v0.3.0
